gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py

```diff
@@ -144,44 +159,41 @@ def _solve(self, rhs, preconditioner=None, num_tridiag=0):
-                            + "not implemented yet."
-                        )
-                    )
-                # in this case we can pull across the diagonals
-                # (\otimes K_i + \otimes D_i) = (\otimes D_i^{1/2})
-                #   (\otimes D_i^{-1/2}K_iD_i^{-1/2} + I)(\otimes D_i^{1/2})
-                # so that
-                # (\otimes K_i + \otimes D_i)^{-1} = (\otimes D_i^{1/2})^{-1}
-                #   \tilde Q (\tilde \Lambda + I)^{-1} \tilde Q (\otimes D_i^{1/2})
-                # Reference: Rakitsch, et al, 2013. "It is all in the noise,"
-                # https://papers.nips.cc/paper/2013/file/59c33016884a62116be975a9bb8257e3-Paper.pdf
-
-                dlt_inv_root, evals_p_i, evecs = _symmetrize_kpadlt_constructor(lt, dlt)
-
-                res1 = evecs._transpose_nonbatch().matmul(dlt_inv_root.matmul(rhs))
-                res2 = evals_p_i.inv_matmul(res1)
-                res3 = evecs.matmul(res2)
-                res = dlt_inv_root.matmul(res3)
+                evals_p_i, evecs = _constant_kpadlt_constructor(lt, dlt)
+                res1 = evals_p_i.inv_matmul(evecs._transpose_nonbatch().matmul(rhs))
+                res = dlt.inv_matmul(evecs.matmul(res1))
                 return res.to(rhs_dtype)
 
-            tmp_term = S.matmul(Lambda_I.inv_matmul(S._transpose_nonbatch().matmul(rhs)))
+            # If the diagonals are not constant, we have to do some more work
+            # since K D^{-1} is generally not symmetric. TODO: implement this solve.
+            if isinstance(lt, KroneckerProductAddedDiagLazyTensor):
+                raise (
+                    NotImplementedError(
+                        "Inverses of KroneckerProductAddedDiagonals and ConstantDiagLazyTensors are "
+                        + "not implemented yet."
+                    )
+                )
+            # in this case we can pull across the diagonals
+            # (\otimes K_i + \otimes D_i) = (\otimes D_i^{1/2})
+            #   (\otimes D_i^{-1/2}K_iD_i^{-1/2} + I)(\otimes D_i^{1/2})
+            # so that
+            # (\otimes K_i + \otimes D_i)^{-1} = (\otimes D_i^{1/2})^{-1}
+            #   \tilde Q (\tilde \Lambda + I)^{-1} \tilde Q (\otimes D_i^{1/2})
+            # Reference: Rakitsch, et al, 2013. "It is all in the noise,"
+            # https://papers.nips.cc/paper/2013/file/59c33016884a62116be975a9bb8257e3-Paper.pdf
+
+            dlt_inv_root, evals_p_i, evecs = _symmetrize_kpadlt_constructor(lt, dlt)
+
+            res1 = evecs._transpose_nonbatch().matmul(dlt_inv_root.matmul(rhs))
+            res2 = evals_p_i.inv_matmul(res1)
+            res3 = evecs.matmul(res2)
+            res = dlt_inv_root.matmul(res3)
+            return res.to(rhs_dtype)
+
+            tmp_term = evecs.matmul(evals_p_i.inv_matmul(evecs._transpose_nonbatch().matmul(rhs)))
             res = lt._inv_matmul(rhs - tmp_term)
 
             return res.to(rhs_dtype)
```

Comment:
something got mixed up here...

Reply:
I'm not following. I renamed the the `S` `Lambda_I` throughout to be more consistent.

Reply:
Oh nevermind i see it.

Reply:
you have two returns in the same indent block...